datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssigner.java

```diff
@@ -5,30 +5,104 @@
-      List<String> assignedPartitions, List<String> unassignedPartitions, int taskCount) {
-    throw new NotImplementedException();
+      Map<String, Set<DatastreamTask>> currentAssignment, List<String> unassignedPartitions,
+      DatastreamGroupPartitionsMetadata partitionMetadata) {
+    String datastreamGroupName = partitionMetadata.getDatastreamGroup().getName();
+    Map<String, PartitionThroughputInfo> partitionInfoMap = throughputInfo.getPartitionInfoMap();
+
+    // filter out all the tasks for the current datastream group, and retain assignments in a map
+    Map<String, Set<String>> newPartitions = new HashMap<>();
+    currentAssignment.values().forEach(tasks -> tasks.forEach(task -> {
+      if (task.getTaskPrefix().equals(datastreamGroupName)) {
+        Set<String> retainedPartitions = new HashSet<>(task.getPartitionsV2());
+        retainedPartitions.retainAll(partitionMetadata.getPartitions());
+        newPartitions.put(task.getId(), retainedPartitions);
+      }
+    }));
+
+    // sort the current tasks on total throughput
+    Map<String, Integer> taskThroughputMap = new HashMap<>();
+    PartitionThroughputInfo defaultPartitionInfo = new PartitionThroughputInfo(DEFAULT_KB_RATE, DEFAULT_MESSAGE_RATE, "");
+    newPartitions.forEach((task, partitions) -> {
+      int totalThroughput = partitions.stream().
+          mapToInt(p -> partitionInfoMap.getOrDefault(p, defaultPartitionInfo).getBytesInKBRate()).sum();
+      taskThroughputMap.put(task, totalThroughput);
+    });
+    List<String> sortedTasks = newPartitions.keySet().stream().sorted(Comparator.comparing(taskThroughputMap::get)).
+        collect(Collectors.toList());
+
+    // sort unassigned partitions on throughput
+    unassignedPartitions.sort((p1, p2) -> {
+      Integer p1KBRate = partitionInfoMap.getOrDefault(p1, defaultPartitionInfo).getBytesInKBRate();
+      Integer p2KBRate = partitionInfoMap.getOrDefault(p2, defaultPartitionInfo).getBytesInKBRate();
+      return p1KBRate.compareTo(p2KBRate);
+    });
+
+    // assign the new partitions one by one
+    while (unassignedPartitions.size() > 0) {
+       String heaviestPartition = unassignedPartitions.remove(unassignedPartitions.size() - 1);
+       int heaviestPartitionThroughput = partitionInfoMap.getOrDefault(heaviestPartition, defaultPartitionInfo)
+           .getBytesInKBRate();
+       String lightestTask = sortedTasks.remove(0);
+       newPartitions.get(lightestTask).add(heaviestPartition);
+       taskThroughputMap.put(lightestTask, taskThroughputMap.get(lightestTask) + heaviestPartitionThroughput);
+       this.insertTaskIntoSortedList(lightestTask, sortedTasks, taskThroughputMap);
+    }
+
+    // build the new assignment using the new partitions for the affected datastream's tasks
+    Map<String, Set<DatastreamTask>> newAssignments = new HashMap<>();
+    currentAssignment.keySet().forEach(instance -> {
+      Set<DatastreamTask> oldTasks = currentAssignment.get(instance);
```

Comment:
Maintaining the sorted list by updating it after each change. Complexity wise it will cost k*log(n) where k is the number of unassigned partitions, n is the number of datastream tasks.